src/GraphQL.Authorization/AuthorizationPolicy.cs

```diff
@@ -1,22 +1,52 @@
-﻿using System.Collections.Generic;
+using System;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace GraphQL.Authorization
 {
-    public interface IAuthorizationPolicy
-    {
-        IEnumerable<IAuthorizationRequirement> Requirements { get; }
-    }
-
+    /// <summary>
+    /// Default implementation for <see cref="IAuthorizationPolicy"/>.
+    /// </summary>
     public class AuthorizationPolicy : IAuthorizationPolicy
     {
+        // allocation optimization for single requirement in policy
+        private readonly IAuthorizationRequirement _singleRequirement;
         private readonly List<IAuthorizationRequirement> _requirements;
 
+        /// <summary>
+        /// Creates a policy with a set of specified requirements.
+        /// </summary>
+        /// <param name="requirements">Specified requirements.</param>
         public AuthorizationPolicy(IEnumerable<IAuthorizationRequirement> requirements)
         {
-            _requirements = new List<IAuthorizationRequirement>(
-                requirements ?? new List<IAuthorizationRequirement>());
+            if (requirements != null)
+            {
+                var temp = requirements.ToList();
+
+                temp.Apply(req => { if (req == null) throw new ArgumentNullException(nameof(requirements), "One of the requirements is null"); });
+
+                if (temp.Count > 1)
+                    _requirements = temp;
+                else if (temp.Count == 1)
+                    _singleRequirement = temp[0];
+            }
         }
 
-        public IEnumerable<IAuthorizationRequirement> Requirements => _requirements;
+        /// <inheritdoc />
+        public IEnumerable<IAuthorizationRequirement> Requirements
+        {
+            get
+            {
+                if (_requirements != null)
+                {
+                    foreach (var r in _requirements)
+                        yield return r;
+                }
+                else if (_singleRequirement != null)
+                {
+                    yield return _singleRequirement;
+                }
+            }
+        }
```

Comment:
```suggestion
            => _requirements ?? (_singleRequirement != null ? new[] { _singleRequirement } : Enumerable.Empty<IAuthorizationRequirement>());
```
This reduces memory allocations in all three cases: for multiple members, for a single member, and for zero members.

However, if this is called on a hot path when it contains a single requirement, it's guaranteed to have lower performance than the ORIGINAL code, because using `yield return` or a single-element array both requires an allocation.  Either way this is in addition to the allocation for the enumerator of the underlying array.  Finally, using `yield return` will prevent other optimizations that attempt to retrieve the `Count` of the list from being able to do so.  (e.g. `new List<>(oldList)`).  So if other code relies on this property, it's nearly guaranteed that this "optimization" will result in poorer performance than the original code here.

Reply:
Yes, I know. I'll revert this code. I wrote it a long time ago and now I am not so happy with it.

Reply:
Reverted.